docs/ovs-offload.md

```diff
@@ -0,0 +1,201 @@
+
+In order to enable Open vSwitch hardware offload, the following steps
+are required. Please make sure you have root privileges to run the commands
+below.
+
+Check the Number of VF Supported on the NIC
+
+```
+cat /sys/class/net/enp3s0f0/device/sriov_totalvfs
+8
+```
+
+Create the VFs
+
+```
+echo '4' > /sys/class/net/enp3s0f0/device/sriov_numvfs
+```
+
+Verfiy that the VFs are created
+
+```
+ip link show enp3s0f0
+8: enp3s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT qlen 1000
+   link/ether a0:36:9f:8f:3f:b8 brd ff:ff:ff:ff:ff:ff
+   vf 0 MAC 00:00:00:00:00:00, spoof checking on, link-state auto
+   vf 1 MAC 00:00:00:00:00:00, spoof checking on, link-state auto
+   vf 2 MAC 00:00:00:00:00:00, spoof checking on, link-state auto
+   vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state auto
+```
+
+Set up the PF to be up
+
+```
+ip link set enp3s0f0 up
+```
+
+Unbind the VFs from the driver
+
+```
+echo 0000:03:00.2 > /sys/bus/pci/drivers/mlx5_core/unbind
+echo 0000:03:00.3 > /sys/bus/pci/drivers/mlx5_core/unbind
+echo 0000:03:00.4 > /sys/bus/pci/drivers/mlx5_core/unbind
+echo 0000:03:00.5 > /sys/bus/pci/drivers/mlx5_core/unbind
+```
+
+Configure SR-IOV VFs to switchdev mode
+
+```
+devlink dev eswitch set pci/0000:03:00.0 mode switchdev
+ethtool -K enp3s0f0 hw-tc-offload on
+```
+
+Bind the VFs to the driver
+
+```
+echo 0000:03:00.2 > /sys/bus/pci/drivers/mlx5_core/bind
+echo 0000:03:00.3 > /sys/bus/pci/drivers/mlx5_core/bind
+echo 0000:03:00.4 > /sys/bus/pci/drivers/mlx5_core/bind
+echo 0000:03:00.5 > /sys/bus/pci/drivers/mlx5_core/bind
+```
+
+## SR-IOV network device plugin configuration
+
+Create a ConfigMap that defines SR-IOV resrouce pool configuration
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: sriovdp-config
+  namespace: kube-system
+data:
+  config.json: |
+    {
+      "resourceList": [{
+          "resourcePrefix": "mellanox.com",
+          "resourceName": "cx5_sriov_switchdev",
+          "isRdma": true,
+          "selectors": {
+                  "vendors": ["15b3"],
+                  "devices": ["1018"],
+                  "drivers": ["mlx5_core"]
+              }
+      }
+      ]
+    }
+```
+
+Deploy SR-IOV network device plugin as DaemonSet see https://github.com/intel/sriov-network-device-plugin
+
+Deploy multus CNI as DaemonSet see https://github.com/intel/multus-cni
+
+Create NetworkAttachementDefinition CRD with Antrea CNI config
+
+```yaml
+Kubernetes Network CRD Spec:
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: mellanox.com/cx5_sriov_switchdev
+spec:
+  config: '{
+    "cniVersion": "0.3.1",
+    "name": "antrea",
+    "plugins": [ { "type": "antrea", "ipam": { "type": "host-local" } }, { "type": "portmap", "capabilities": {"portMappings": true} }]
+}'
+```
+## Deploy Antrea Image with hw-offload enabled
```

Comment:
I am thinking might be better to add a config option in Antrea agent about HW offload enabled, and Agent will validate the option to accept the VF deviceID from CNI. And if we do that the start_ovs script can read the option too.
This way seems more explicit regarding to the OVS and Agent configuration.
What you guys think? @antoninbas

Reply:
This will limited you as you won't be able to deploy veth pods and SR-IOV pods in the same cluster.  SR-IOV VF limited to 64 per port. So the maybe case when you still want veth POD and SR-IOV pod in the same cluster

Reply:
Got it. But even we add a config option, it does not mean all Pods must use SR-IOV. It just means the feature is enabled (and implicitly means OVS must have been configured to support it).

Reply:
yes, I can add config option and it that case it will validate that ovs is configure with hw-offload=true flag

Reply:
That sounds good to me.

Reply:
@jianjuns are you sure about that extra config knob? It seems like one more thing the user has to do:
1) change the command for `start_ovs`
2) set the config parameter

I think originally you meant to have the second one replace the first one:
> I am thinking might be better to add a config option in Antrea agent about HW offload enabled, and Agent will validate the option to accept the VF deviceID from CNI. And if we do that the start_ovs script can read the option too.

But that means that the Antrea Agent ConfigMap now needs to be mounted in the antrea-ovs container, and the JSON conf needs to be parsed there to extract the value. Is that what you had in mind?

Reply:
Humm.. I thought antrea-ovs already mounts the configmap and reads the bridge name option from it, but seems that was when we used start_ovs to create the bridge.

Do we have another way in Agent to check OVS is configured correctly to support offload?
I just feel CNI deviceID parameter is too weak a contract, and want to see if we have other ways to validate the request.

Reply:
@moshe010 in order to make progress, what is the current behavior if one tries to attach an SRIOV device to an OVS bridge, without setting `hw-offload=true` when starting vswitchd? Does it fail, and if yes, how?

Reply:
Do we have a way to check in antrea-agent that hw-offload is enabled on OVS?

Reply:
only checking ovs hw-offload in other config

Reply:
Sorry, I did not get your reply. My question is whether we have a way to validate that hw-offload is enabled for the OVS bridge in the antrea-agent (e.g. checking any OVS bridge attribute at antrea-agent startup), and so antrea-agent can check it when receiving a CNI ADD with deviceID, and might just fail the call if hw-offload is not enabled, instead of blindly accept the call.

Reply:
So on the cmd add we can check if the hw-offload=true if get DeviceID. Is work for you?

Reply:
Ok. I now got hw-offload=true is an OVS config. Sounds good to me then?